### ISSUE TITLE
dhall freeze --verify

### ISSUE THREAD

**Issue by Profpatsch** [dhall-lang/dhall-haskell#1629]
Verify that the hashes of frozen inputs are the ones described by the expressions they freeze. This is to prevent inconsistencies like the one mentioned in [this issue](https://github.com/dhall-lang/dhall-lang/issues/666#issuecomment-574114108):

```
missing sha256:…
```

where someone changes the expression to something else and it “works on my machine”, because the cache still has a reference to the hash.

People can add `dhall freeze --verify` invocations to CI, to check for such problems automatically.

**Comment by Gabriel439** [comment]
@Profpatsch: `dhall freeze` already ignores the cache.  If you give `dhall freeze` an import with an integrity check it will ignore the integrity check and refetch the import and recompute what the integrity check should be.

For example:

```bash
$ dhall freeze <<< 'https://example.com/ sha256:ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff as Text'
https://example.com/ sha256:fc4912b5b5191458d7cb0e1ca16dcf233c654c95fba14b4351f17cfbd59a2853 as Text
```

**Comment by ari-becker** [comment]
This would be great, it would be easy to run in parallel with the normal build such that the normal build benefits from the presence of the cache and a separate task verifies the origins behind the cache. Therefore this has nominal effect on build times while adding confidence that a green CI build means that the code is still buildable on a cold/empty cache.

**Comment by ari-becker** [comment]
@Gabriel439 fwiw, I think `dhall freeze --verify --file foo.bar` is more discoverable / clear than `diff <(dhall freeze --file foo.bar) <(dhall --file foo.bar)`

**Comment by Profpatsch** [comment]
yeah, just a shortcut for the above. But worthwhile I think. Can also show way better diffs than `diff` could, e.g. how exactly the expression in the file differs from the expression in the cache.

**Comment by Gabriel439** [comment]
@Profpatsch: Note that replacing `diff` with `dhall diff` can do that, too, but I plan to implement this anyway

**Comment by Gabriel439** [comment]
Note that I will propose to call this flag `--check` (for consistency with `dhall format`)

**Comment by Gabriel439** [comment]
The fix is up here: https://github.com/dhall-lang/dhall-haskell/pull/1636
